js/ui/mouse.ts

```diff
@@ -1,30 +1,84 @@
 import type Mouse from '../cards/mouse';
 import { enableMouseMode } from './joystick';
 
+type TouchEventTarget = TouchEvent & { target: HTMLCanvasElement };
+
 export class MouseUI {
     private mouse: Mouse;
 
     constructor(private canvas: HTMLCanvasElement)  {
-        this.canvas.addEventListener(
-            'mousemove',
-            (event: MouseEvent & { target: HTMLCanvasElement} ) => {
-                const { offsetX, offsetY, target } = event;
-                this.mouse.setMouseXY(
-                    offsetX,
-                    offsetY,
-                    target.clientWidth,
-                    target.clientHeight
-                );
+        const updateTouchXY = (event: TouchEventTarget) => {
+            const { targetTouches, target } = event;
+            if (targetTouches.length < 1) {
+                return;
             }
-        );
+            const rect = target.getBoundingClientRect();
+            const leftPad = (rect.width - 560) / 2 + rect.left;
+            const topPad = (rect.height - 384) / 2 + rect.top;
+            const { clientX, clientY } = targetTouches[0];
+            const xPos = clientX - leftPad;
+            const yPos = clientY - topPad;
+            this.mouse.setMouseXY(
+                Math.max(Math.min(xPos, 559), 0),
+                Math.max(Math.min(yPos, 383), 0),
+                560,
+                384,
+            );
+        };
+
+        if ('ontouchstart' in window) {
+            this.canvas.addEventListener(
+                'touchmove',
+                (event: TouchEventTarget) => {
+                    updateTouchXY(event);
+                }
+            );
 
-        this.canvas.addEventListener('mousedown', () => {
-            this.mouse.setMouseDown(true);
-        });
+            this.canvas.addEventListener(
+                'touchstart',
+                (event: TouchEventTarget) => {
+                    updateTouchXY(event);
+                    setTimeout(() => this.mouse.setMouseDown(true), 100);
```

Comment:
Why the timeout? It seems like it could make the mouse down action happen at a different point if `touchmove` happens in between.

Reply:
This was to handle a behavior in Apple II Desktop where it processes clicks before movement, so if a click and a move happen at the same time, which is most tap interactions, it was treated as a click then drag. So the timeout is to make sure they get handled separate, and I hope is short enough to avoid other issues. I'll add a comment.